src/interfaces/controllers/deregister-trailer/deregister-trailer.ts

```diff
@@ -0,0 +1,54 @@
+import { Request, Response, NextFunction } from 'express';
+import * as domain from '../../../domain';
+import { DataAccess } from '../../../utils/data-access';
+import { IRequestHandler } from '../i-request-handler';
+import { deregisterTrailerValidator } from './deregister-trailer-validator';
+
+export class DeregisterTrailer implements IRequestHandler {
+  private validate(payload?: domain.DeregisterTrailer): string {
+    const errors = deregisterTrailerValidator.validate(payload, { abortEarly: false }).error;
+    return errors?.message;
+  }
+
+  public async call(req: Request, res: Response, next: NextFunction): Promise<void> {
+    const deregisterTrailer = req.body as domain.DeregisterTrailer;
+    const { trn } = req.params;
+
+    const errors = this.validate(deregisterTrailer);
+    if (errors) {
+      next(new domain.HTTPError(400, errors));
+      return;
+    }
+
+    try {
+      const { reasonForDeregistration, deregisterDate } = deregisterTrailer;
+      let existingTrailerRegistration = await this.checkForExistingTrailerRegistration(trn);
+      if (existingTrailerRegistration) {
+        existingTrailerRegistration = {
+          ...existingTrailerRegistration,
+          reasonForDeregistration,
+          deregisterDate,
+        };
+        await DataAccess.getInstance().upsertTrailerRegistration(existingTrailerRegistration);
+      }
+    } catch (error) {
+      next(new domain.HTTPError(500, error));
+      return;
+    }
+    res.status(204).send();
+  }
+
+  private async checkForExistingTrailerRegistration(trn: string): Promise<domain.TrailerRegistration | null> {
+    const trailerRegistration = await DataAccess.getInstance().getByTrn(trn);
+    if (!trailerRegistration) {
```

Comment:
might need some clarification on the requirements or we will need to discuss if a use case is possible and what we do about it with @bristol-mike-jones 

```
AC4 - Unregister when TRN is found

GIVEN I am a DVLA API user

WHEN I make a PUT request to /v1/trailers/deregister/{trn} with a valid deregister payload

AND the TRN does exist, and is not already deregistered

THEN the TRN is updated with unregister payload in the database

AND I get a HTTP 204 response code


```

Can we have the case where

the TRN does exist, and **is** already deregistered ?

Reply:
@bristol-mike-jones please provide your feedback.

Reply:
Add logs to hi light such scenario.